clientv3: handle nil connection after *Client.Close (KV)

gyuho opened the pull request:
For https://github.com/coreos/etcd/issues/5495.

This only handles KV operations for now.

Need to discuss how to handle this for watch operations.

heyitsanthony commented:
Fixes the crash for me. Add a clientv3/integration test similar to https://github.com/coreos/etcd/issues/5495#issuecomment-222570871 too?

gyuho commented:
Yeah will add more tests today. Thanks!

gyuho commented:
@heyitsanthony PTAL. Will address watch part in a separate PR.

Thanks!

heyitsanthony commented on integration/cluster.go:
```diff
@@ -769,11 +779,24 @@ func NewClusterV3(t *testing.T, cfg *ClusterConfig) *ClusterV3 {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for i, client := range c.clients {
+		if _, ok := c.closedClientIdx[i]; ok {
```
instead of a map, can this set the closed client to nil/check if the `client` is nil before closing?

gyuho replied:
Right. That's the better way. Will fix. Thanks.

heyitsanthony commented on clientv3/integration/kv_test.go:
```diff
@@ -279,6 +279,42 @@ func TestKVRange(t *testing.T) {
+	if err := cli.Close(); err != nil {
+		t.Fatal(err)
+	}
+	clus.SetClosedClient(0)
```
`TakeClient(0)`?

gyuho commented:
@heyitsanthony PTAL. Thanks.

Test failure on Watch seems un-related.

```
=== RUN   TestWatchReconnRequest
2016-05-31 18:48:10.260992 I | integration: launching node6394570431849258471 (unix://localhost:node6394570431849258471.sock)
2016-05-31 18:48:10.281859 I | etcdserver: name = node6394570431849258471
2016-05-31 18:48:10.281903 I | etcdserver: data dir = /tmp/etcd250924617
2016-05-31 18:48:10.281930 I | etcdserver: member dir = /tmp/etcd250924617/member
2016-05-31 18:48:10.281956 I | etcdserver: heartbeat = 10ms
2016-05-31 18:48:10.281977 I | etcdserver: election = 100ms
2016-05-31 18:48:10.281996 I | etcdserver: snapshot count = 0
2016-05-31 18:48:10.282021 I | etcdserver: advertise client URLs = http://127.0.0.1:21152
2016-05-31 18:48:10.282047 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21151
2016-05-31 18:48:10.282092 I | etcdserver: initial cluster = node2464910841826418901=http://127.0.0.1:21153,node2484537562608410507=http://127.0.0.1:21149,node6394570431849258471=http://127.0.0.1:21151
2016-05-31 18:48:10.284495 I | etcdserver: name = node2484537562608410507
2016-05-31 18:48:10.284528 I | etcdserver: data dir = /tmp/etcd188198418
2016-05-31 18:48:10.284556 I | etcdserver: member dir = /tmp/etcd188198418/member
2016-05-31 18:48:10.284580 I | etcdserver: heartbeat = 10ms
2016-05-31 18:48:10.284599 I | etcdserver: election = 100ms
2016-05-31 18:48:10.284618 I | etcdserver: snapshot count = 0
2016-05-31 18:48:10.284641 I | etcdserver: advertise client URLs = http://127.0.0.1:21150
2016-05-31 18:48:10.284664 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21149
2016-05-31 18:48:10.284706 I | etcdserver: initial cluster = node2464910841826418901=http://127.0.0.1:21153,node2484537562608410507=http://127.0.0.1:21149,node6394570431849258471=http://127.0.0.1:21151
2016-05-31 18:48:10.285178 I | etcdserver: name = node2464910841826418901
2016-05-31 18:48:10.285221 I | etcdserver: data dir = /tmp/etcd670417172
2016-05-31 18:48:10.285262 I | etcdserver: member dir = /tmp/etcd670417172/member
2016-05-31 18:48:10.285293 I | etcdserver: heartbeat = 10ms
2016-05-31 18:48:10.285321 I | etcdserver: election = 100ms
2016-05-31 18:48:10.285350 I | etcdserver: snapshot count = 0
2016-05-31 18:48:10.285386 I | etcdserver: advertise client URLs = http://127.0.0.1:21154
2016-05-31 18:48:10.285421 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21153
2016-05-31 18:48:10.285494 I | etcdserver: initial cluster = node2464910841826418901=http://127.0.0.1:21153,node2484537562608410507=http://127.0.0.1:21149,node6394570431849258471=http://127.0.0.1:21151
2016-05-31 18:48:10.296517 I | etcdserver: starting member dc3cde9072b18b64 in cluster 99014a85ff3f6e53
2016-05-31 18:48:10.296666 I | raft: dc3cde9072b18b64 became follower at term 0
2016-05-31 18:48:10.296718 I | raft: newRaft dc3cde9072b18b64 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-31 18:48:10.296750 I | raft: dc3cde9072b18b64 became follower at term 1
2016-05-31 18:48:10.307087 I | etcdserver: set snapshot count to default 10000
2016-05-31 18:48:10.307173 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-31 18:48:10.314918 N | etcdserver: added member 61c8b1a6fb4577e2 [http://127.0.0.1:21149] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.324172 I | raft: 61c8b1a6fb4577e2 became follower at term 0
2016-05-31 18:48:10.324241 I | raft: newRaft 61c8b1a6fb4577e2 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-31 18:48:10.315792 I | integration: launched node6394570431849258471 (unix://localhost:node6394570431849258471.sock)
2016-05-31 18:48:10.316066 N | etcdserver: added member c08326d114e345fa [http://127.0.0.1:21153] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.317723 N | etcdserver: added local member dc3cde9072b18b64 [http://127.0.0.1:21151] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.323986 I | etcdserver: starting member 61c8b1a6fb4577e2 in cluster 99014a85ff3f6e53
2016-05-31 18:48:10.331075 I | raft: c08326d114e345fa became follower at term 0
2016-05-31 18:48:10.331137 I | raft: newRaft c08326d114e345fa [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-31 18:48:10.331175 I | raft: c08326d114e345fa became follower at term 1
2016-05-31 18:48:10.372406 I | etcdserver: set snapshot count to default 10000
2016-05-31 18:48:10.324293 I | raft: 61c8b1a6fb4577e2 became follower at term 1
2016-05-31 18:48:10.330914 I | etcdserver: starting member c08326d114e345fa in cluster 99014a85ff3f6e53
2016-05-31 18:48:10.372471 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-31 18:48:10.376672 I | rafthttp: the connection with dc3cde9072b18b64 became active
2016-05-31 18:48:10.412329 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-31 18:48:10.414848 I | integration: launched node2484537562608410507 (unix://localhost:node2484537562608410507.sock)
2016-05-31 18:48:10.408506 I | raft: dc3cde9072b18b64 is starting a new election at term 1
2016-05-31 18:48:10.408586 I | raft: dc3cde9072b18b64 became candidate at term 2
2016-05-31 18:48:10.408620 I | raft: dc3cde9072b18b64 received vote from dc3cde9072b18b64 at term 2
2016-05-31 18:48:10.408663 I | raft: dc3cde9072b18b64 [logterm: 1, index: 3] sent vote request to 61c8b1a6fb4577e2 at term 2
2016-05-31 18:48:10.408704 I | raft: dc3cde9072b18b64 [logterm: 1, index: 3] sent vote request to c08326d114e345fa at term 2
2016-05-31 18:48:10.412260 I | etcdserver: set snapshot count to default 10000
2016-05-31 18:48:10.422566 N | etcdserver: added local member 61c8b1a6fb4577e2 [http://127.0.0.1:21149] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.422969 N | etcdserver: added member c08326d114e345fa [http://127.0.0.1:21153] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.418695 I | rafthttp: the connection with dc3cde9072b18b64 became active
2016-05-31 18:48:10.419808 I | rafthttp: the connection with 61c8b1a6fb4577e2 became active
2016-05-31 18:48:10.425383 N | etcdserver: added member 61c8b1a6fb4577e2 [http://127.0.0.1:21149] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.425788 N | etcdserver: added member dc3cde9072b18b64 [http://127.0.0.1:21151] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.426164 N | etcdserver: added local member c08326d114e345fa [http://127.0.0.1:21153] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.426735 N | etcdserver: added member dc3cde9072b18b64 [http://127.0.0.1:21151] to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.427515 I | integration: launched node2464910841826418901 (unix://localhost:node2464910841826418901.sock)
2016-05-31 18:48:10.432045 I | rafthttp: the connection with c08326d114e345fa became active
2016-05-31 18:48:10.437455 I | rafthttp: the connection with c08326d114e345fa became active
2016-05-31 18:48:10.437981 I | rafthttp: the connection with 61c8b1a6fb4577e2 became active
2016-05-31 18:48:10.438755 I | raft: 61c8b1a6fb4577e2 is starting a new election at term 1
2016-05-31 18:48:10.438855 I | raft: 61c8b1a6fb4577e2 became candidate at term 2
2016-05-31 18:48:10.438911 I | raft: 61c8b1a6fb4577e2 received vote from 61c8b1a6fb4577e2 at term 2
2016-05-31 18:48:10.441698 I | raft: c08326d114e345fa [logterm: 1, index: 3, vote: 0] voted for 61c8b1a6fb4577e2 [logterm: 1, index: 3] at term 2
2016-05-31 18:48:10.442746 I | raft: c08326d114e345fa [logterm: 1, index: 3, vote: 61c8b1a6fb4577e2] rejected vote from dc3cde9072b18b64 [logterm: 1, index: 3] at term 2
2016-05-31 18:48:10.441506 I | raft: c08326d114e345fa [term: 1] received a MsgVote message with higher term from 61c8b1a6fb4577e2 [term: 2]
2016-05-31 18:48:10.441612 I | raft: c08326d114e345fa became follower at term 2
2016-05-31 18:48:10.438976 I | raft: 61c8b1a6fb4577e2 [logterm: 1, index: 3] sent vote request to c08326d114e345fa at term 2
2016-05-31 18:48:10.439029 I | raft: 61c8b1a6fb4577e2 [logterm: 1, index: 3] sent vote request to dc3cde9072b18b64 at term 2
2016-05-31 18:48:10.443414 I | raft: dc3cde9072b18b64 [logterm: 1, index: 3, vote: dc3cde9072b18b64] rejected vote from 61c8b1a6fb4577e2 [logterm: 1, index: 3] at term 2
2016-05-31 18:48:10.444829 I | raft: 61c8b1a6fb4577e2 [logterm: 1, index: 3, vote: 61c8b1a6fb4577e2] rejected vote from dc3cde9072b18b64 [logterm: 1, index: 3] at term 2
2016-05-31 18:48:10.445640 I | raft: dc3cde9072b18b64 received vote rejection from 61c8b1a6fb4577e2 at term 2
2016-05-31 18:48:10.445697 I | raft: dc3cde9072b18b64 [quorum:2] has received 1 votes and 1 vote rejections
2016-05-31 18:48:10.446299 I | raft: dc3cde9072b18b64 received vote rejection from c08326d114e345fa at term 2
2016-05-31 18:48:10.446352 I | raft: dc3cde9072b18b64 [quorum:2] has received 1 votes and 2 vote rejections
2016-05-31 18:48:10.446419 I | raft: dc3cde9072b18b64 became follower at term 2
2016-05-31 18:48:10.446591 I | raft: 61c8b1a6fb4577e2 became leader at term 2
2016-05-31 18:48:10.446635 I | raft: raft.node: 61c8b1a6fb4577e2 elected leader 61c8b1a6fb4577e2 at term 2
2016-05-31 18:48:10.446498 I | raft: 61c8b1a6fb4577e2 received vote from c08326d114e345fa at term 2
2016-05-31 18:48:10.446540 I | raft: 61c8b1a6fb4577e2 [quorum:2] has received 2 votes and 0 vote rejections
2016-05-31 18:48:10.447278 I | raft: raft.node: c08326d114e345fa elected leader 61c8b1a6fb4577e2 at term 2
2016-05-31 18:48:10.447610 I | raft: raft.node: dc3cde9072b18b64 elected leader 61c8b1a6fb4577e2 at term 2
2016-05-31 18:48:10.451765 I | etcdserver: published {Name:node2484537562608410507 ClientURLs:[http://127.0.0.1:21150]} to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.454767 I | etcdserver: published {Name:node6394570431849258471 ClientURLs:[http://127.0.0.1:21152]} to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.454957 I | etcdserver: published {Name:node2464910841826418901 ClientURLs:[http://127.0.0.1:21154]} to cluster 99014a85ff3f6e53
2016-05-31 18:48:10.457092 I | etcdserver: setting up the initial cluster version to 3.0
2016-05-31 18:48:10.459906 N | membership: set the initial cluster version to 3.0
2016-05-31 18:48:10.460501 N | membership: set the initial cluster version to 3.0
2016-05-31 18:48:10.460983 N | membership: set the initial cluster version to 3.0
panic: send on closed channel

goroutine 12248 [running]:
panic(0xf8c3c0, 0xc825d78360)
	/usr/local/golang/1.6.2/go/src/runtime/panic.go:481 +0x3ff
github.com/coreos/etcd/etcdserver/api/v3rpc.(*serverWatchStream).recvLoop(0xc823b51650, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:180 +0xb6a
github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch.func2(0xc82063a420, 	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:126 +0x2fc
0xc823b51650)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:126 +0x2f
created by github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch
exit status 2
FAIL	github.com/coreos/etcd/clientv3/integration	38.493s
```

heyitsanthony commented on integration/cluster.go:
```diff
@@ -769,12 +774,25 @@ func NewClusterV3(t *testing.T, cfg *ClusterConfig) *ClusterV3 {
 
+func (c *ClusterV3) TakeClient(idx int) error {
+	c.mu.Lock()
+	err := c.clients[idx].Close()
```
don't close here-- that way the caller can take ownership of the client from the integration cluster

gyuho replied:
ok will fix and rerun tests. Thanks!

heyitsanthony commented:
lgtm

gyuho commented:
Thanks!
